src/wx/background-input.cpp

```diff
@@ -0,0 +1,2331 @@
+    }
+
+    x11display = XOpenDisplay(NULL);
+    if (x11display == NULL) {
+        return (wxThread::ExitCode)1;
+    }
+    XQueryKeymap(x11display, previousState);
+#endif
+    return (wxThread::ExitCode)0;
+}
+
+wxThread::ExitCode BackgroundInput::CheckKeyboard()
+{
+#if defined(__WXMSW__)
+    for (int i = 0x08; i < 0xFF; ++i) {
+        SHORT bits = GetAsyncKeyState(i);
+        if (bits == previousState[i]) continue;
+        wchar_t uc = 0;
+        int xKeySym = VKToWX(i, 0, &uc);
+        // virtual key "i" is pressed
+        if ((bits & 0x8000) && (previousState[i] & 0x8000) == 0) {
+            if (handler && !wxWindow::FindFocus()) {
+                wxKeyEvent ev(wxEVT_KEY_DOWN);
+                ev.m_keyCode = xKeySym;
+                handler->AddPendingEvent(ev);
+            }
+        }
+        // virtual key "i" is released
+        else if (((bits & 0x8000) == 0) && (previousState[i] & 0x8000)) {
+            if (handler && !wxWindow::FindFocus()) {
+                wxKeyEvent ev(wxEVT_KEY_UP);
+                ev.m_keyCode = xKeySym;
+                handler->AddPendingEvent(ev);
+            }
+        }
+        previousState[i] = bits;
+    }
+#elif defined(__WXMAC__)
+#else // defined(__WXGTK__)
+    XQueryKeymap(x11display, currentState);
+    for (int i = 0; i < 32; ++i) {
+        int cte = 8*i;
+        if (currentState[i] != previousState[i]) {
+            int numCurrent = currentState[i];
+            int numPrevious = previousState[i];
+            for (int pos = 0; pos < 8; ++pos) {
+                int rawKeycode = cte + pos;
+                int xKeySym = wxUnicodeCharXToWX(XkbKeycodeToKeysym(x11display, rawKeycode, 0, 1)); // same as WX keycode wxEVT_KEY_DOWN
+                // Key was pressed
+                if (((numCurrent & 0x01) == 1) && ((numPrevious & 0x01) == 0)) {
+                    if (handler && !wxWindow::FindFocus()) {
+                        wxKeyEvent ev(wxEVT_KEY_DOWN);
+                        ev.m_keyCode = xKeySym;
+                        handler->AddPendingEvent(ev);
+                    }
+                }
+                // Key was released
+                else if (((numCurrent & 0x01) == 0) && ((numPrevious & 0x01) == 1)) {
+                    if (handler && !wxWindow::FindFocus()) {
+                        wxKeyEvent ev(wxEVT_KEY_UP);
+                        ev.m_keyCode = xKeySym;
+                        handler->AddPendingEvent(ev);
+                    }
+                }
+                numCurrent /= 2;
+                numPrevious /= 2;
+            }
+        }
+        previousState[i] = currentState[i];
+    }
+#endif
+    return (wxThread::ExitCode)0;
+}
+
+void BackgroundInput::Cleanup()
+{
+#if defined(__WXMSW__)
+#elif defined(__WXMAC__)
+#else // defined(__WXGTK__)
+    if (x11display) {
+        XCloseDisplay(x11display);
+        x11display = nullptr;
+    }
+#endif
+}
+
+wxThread::ExitCode BackgroundInput::Entry()
+{
+    if (Setup())
```

Comment:
Minor nitpick, feel free to ignore, I would write this as:

```cpp
if (Setup() != 0)
```

otherwise it looks like you are testing a boolean, which makes it more confusing.